File: April-Challenges-2021/Day-3/Longest-Valid-Parentheses.cpp
class Solution {
public:
    int longestValidParentheses(string s) {
        vector<int>stack;
        stack.push_back(-1);
        int max_pars = 0;
        for(int i = 0; i<s.size(); i++){
            if(s[i]=='('){
                stack.push_back(i);
            }else{
                stack.pop_back();
                if (stack.size()==0){
                    stack.push_back(i);
                }else{
                    max_pars = max(max_pars,i-stack.back());
                }
            }
        }
        return max_pars;
    }
};

File: April-Challenges-2021/Day-3/Longest-Valid-Parentheses.py
 
class Solution:
    def longestValidParentheses(self, s: str) -> int:
        max_pars = 0
        stack = []
        stack.append(-1)
        for i in range(len(s)):
            if s[i] == '(':
                stack.append(i)
            else:
                stack.pop()
                if len(stack)==0:
                    stack.append(i)
                else:
                    max_pars = max(max_pars, i-stack[-1])
        return max_pars     
